build_tools/python/e2e_test_framework/definitions/utils.py

```diff
@@ -0,0 +1,50 @@
+## Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""Utils that help construct definitions."""
+
+from typing import Callable, List, Sequence
+
+MAX_SUBSTITUTION_ITERATIONS = 10
+
+
+def transform_flags(flags: Sequence[str],
+                    map_funcs: Sequence[Callable[[str], str]]) -> List[str]:
+  """Call map functions to transform flag values, e.g., replace placeholders
+  that were unknown when the flag was constructed.
+
+  It parses and extracts the flag values from both keyword and positional flags,
+  transforms them, and returns the updated flags with transformed values.
+
+  Each flag value is transformed only once by each map function in order.
+  
+  Args:
+    flags: list of flags.
+    map_funcs: list of map functions to map flag value.
+  Returns:
+    List of transformed flags.
+  """
+
+  transformed_flags = []
+  for flag in flags:
+    if flag.startswith("-"):
+      # Keyword argument
+      split_pos = flag.find("=")
+      # Do nothing if there is no flag value.
+      if split_pos == -1:
+        transformed_flags.append(flag)
+        continue
+      value_pos = split_pos + 1
+    else:
+      # Positional argument
+      value_pos = 0
+
+    new_value = flag[value_pos:]
+    for map_func in map_funcs:
+      new_value = map_func(new_value)
+
+    transformed_flags.append(flag[:value_pos] + new_value)
```

Comment:
I think this can be a lot cleaner with [partition](https://docs.python.org/3/library/stdtypes.html#str.partition)

```suggestion
    keyword, separator, value = ("", "", flag)
    if flag.startswith("-"):
      keyword, separator, value = flag.partition("=")    

    if value:
      for map_func in map_funcs:
        value = map_func(value)

    transformed_flags.append(f"{keyword}{separator}{value}")
```